### PR TITLE
Fix file descriptor leak in repo version repair

### PR DESCRIPTION
**Pull request by dralley**
closes: #7735
https://pulp.plan.io/issues/7735

### CODE REVIEW

File: CHANGES/7735.bugfix [hunk 1]
```diff
@@ -0,0 +1 @@
+Fix a file descriptor leak during repository version repair operations.
```
Comment (ipanova) [inline CHANGES/7735.bugfix:1]: s/Fix/Fixed

File: pulpcore/app/tasks/repository.py [hunk 2]
```diff
@@ -84,8 +84,10 @@ async def _repair_repository_version(version):
         ) as repaired:
             query_set = models.ContentArtifact.objects.filter(
                 content__in=version.content
-            ).prefetch_related("artifact")
-            for content_artifact in query_set:
+            ).select_related(
+                "artifact"
+            )
+            for content_artifact in query_set.iterator():
```
Comment (ipanova) [inline pulpcore/app/tasks/repository.py:90]: :+1

File: pulpcore/app/tasks/repository.py [hunk 3]
```diff
@@ -58,12 +58,16 @@ async def _repair_ca(content_artifact, repaired=None):
...
+
+
 def _verify_ca(content_artifact):
     try:
         # verify files digest
         hasher = hashlib.sha256()
-        for chunk in iter(lambda: content_artifact.artifact.file.read(1024 * 1024), b""):
+        for chunk in content_artifact.artifact.file.chunks(CHUNK_SIZE):
             hasher.update(chunk)
+        content_artifact.artifact.file.close()
```
Comment (dralley) [inline pulpcore/app/tasks/repository.py:70]: The "actual fix"

**Review by ipanova — APPROVED** [review]

**Review by dkliban — APPROVED** [review]

File: pulpcore/app/tasks/repository.py [hunk 4]
```diff
@@ -80,8 +85,8 @@ async def _repair_repository_version(version):
         ) as repaired:
             query_set = models.ContentArtifact.objects.filter(
                 content__in=version.content
-            ).prefetch_related("artifact")
-            for content_artifact in query_set:
+            ).select_related("artifact")
+            for content_artifact in query_set.iterator():
```
Comment (goosemania) [inline pulpcore/app/tasks/repository.py:89]: This should improve memory usage, right?
I suggest to make such improvements visible through the changelog.
Not insisting to do it for this one. Just keep in mind that calling out any perf/memory improvements in our blog or release notes is a good advertisement of the fact that we care about the scalability and quality of the project.

Comment (dralley) [reply pulpcore/app/tasks/repository.py:89]: Probably, but not by a huge amount.  ContentArtifacts are pretty small and we're dealing with, at most, however many of them are present in the RepositoryVersion.

My general guideline is that I don't want to advertise "probablies" that weren't measured, and I don't really think it's worth measuring, and the quality will speak for itself long-term.

File: pulpcore/app/tasks/repository.py [hunk 5]
```diff
@@ -58,12 +59,16 @@ async def _repair_ca(content_artifact, repaired=None):
...
 
+CHUNK_SIZE = 1024 * 1024  # 1 Mb
+
+
 def _verify_ca(content_artifact):
     try:
         # verify files digest
         hasher = hashlib.sha256()
-        for chunk in iter(lambda: content_artifact.artifact.file.read(1024 * 1024), b""):
+        for chunk in content_artifact.artifact.file.chunks(CHUNK_SIZE):
```
Comment (mdellweg) [inline pulpcore/app/tasks/repository.py:69]: :1st_place_medal:

File: pulpcore/app/tasks/repository.py (hunk 4 above)
Comment (goosemania) [reply pulpcore/app/tasks/repository.py:89]: I agree with the time-will-tell approach.
I just think that these are different target audiences we are talking about :) , people who are users already and will be happy with time that pulp is getting better and notice performance improvements, and people who look at projects and evaluate which one to choose. The latter ones potentially skim through the changelog or blog posts.

**Comment by ekohl** [comment]
I just ran into this, but isn't there a `with` block equivalent? if there's an exception the file still isn't closed. While a bit of an edge case, it's still good to cover IMHO.

**Comment by dralley** [comment]
@ekohl Well, I'm not sure if it could possibly throw an exception other than `FileNotFound`, but at least in that case there would be no open file.

**Comment by ekohl** [comment]
Never assume it won't fail? I think this actually works and is safer:
```python
with content_artifact.artifact.file as fp:
    for chunk in fp.chunks(CHUNK_SIZE):
        hasher.update(chunk)
```

**Comment by dralley** [comment]
I don't believe these objects work in that manner

**Comment by ekohl** [comment]
`django.core.files.File` does implement the protocol:
https://github.com/django/django/blob/225d96533a8e05debd402a2bfe566487cc27d95f/django/core/files/base.py#L102-L106
(I linked latest but in 2.2 it's the same)

**Pull request merged**